### PR TITLE
Consolidate Render(Ui)Materials(2d) into RenderAssets

### PR DESCRIPTION
**Pull request by superdump**
# Objective

- Replace `RenderMaterials` / `RenderMaterials2d` / `RenderUiMaterials` with `RenderAssets` to enable implementing changes to one thing, `RenderAssets`, that applies to all use cases rather than duplicating changes everywhere for multiple things that should be one thing.
- Adopts #8149 

## Solution

- Make RenderAsset generic over the destination type rather than the source type as in #8149 
- Use `RenderAssets<PreparedMaterial<M>>` etc for render materials

---

## Changelog

- Changed:
  - The `RenderAsset` trait is now implemented on the destination type. Its `SourceAsset` associated type refers to the type of the source asset.
  - `RenderMaterials`, `RenderMaterials2d`, and `RenderUiMaterials` have been replaced by `RenderAssets<PreparedMaterial<M>>` and similar.

## Migration Guide

- `RenderAsset` is now implemented for the destination type rather that the source asset type. The source asset type is now the `RenderAsset` trait's `SourceAsset` associated type.

### CODE REVIEW

**Comment by superdump** [comment]
I would love to find a way to not require people to implement `AssetUsages` for custom materials, and instead just use a default. But I haven't had any good ideas yet. I can't implement for AssetUsages for T at least.

File: crates/bevy_render/src/render_asset.rs [hunk 1]
```diff
@@ -19,32 +19,34 @@ pub enum PrepareAssetError<E: Send + Sync + 'static> {
     RetryNextUpdate(E),
 }
 
+pub trait AssetUsages {
```
Comment (JMS55) [inline crates/bevy_render/src/render_asset.rs:22]: I don't love the naming on this tbh :/

Comment (superdump) [reply crates/bevy_render/src/render_asset.rs:22]: Me neither. I can't call it `RenderAssetUsages` though, as that is already taken.

Comment (JMS55) [reply crates/bevy_render/src/render_asset.rs:22]: Perhaps remove the enum, and make the RenderAssetUsages trait have two methods, `main_world() -> bool` and `render_world() -> bool`?

**Comment by superdump** [comment]
@infmagic2047 great idea! So simple! :) I love it.

File: crates/bevy_render/src/render_asset.rs (hunk 1 above)
Comment (superdump) [reply crates/bevy_render/src/render_asset.rs]: It is now gone. :)

**Review by pcwalton — APPROVED** [review]
This looks fine. I'll trust you that it was necessary to move `RenderAsset` over to be implemented on the render world type.

**Comment by IceSentry** [comment]
@pcwalton here's the justification from the original PR: 

> Remove one to one correlation between "main world" asset and GPU asset, by swapping base type that implements RenderAsset to be the GPU asset type. That would make RenderAssetPlugin more flexible, being able to prepare different data based on same source asset (example: effect plugin could preprocess mesh buffer data for specialized rendering, without affecting main render flow).
That also makes RenderAssets<GPUType> more explicit.

**Review by IceSentry — APPROVED** [review]

**Comment by superdump** [comment]
Please don’t merge this yet. I want to think about a couple of things with it so I am convinced implementing RenderAsset on the destination type is better.

**Comment by superdump** [comment]
I turned out that robtfm had a branch that did basically the same thing last year after I commented that implementing RenderAsset on types that implement Materials was problematic in various ways:
* Problems I discovered: https://discord.com/channels/691052431525675048/743663924229963868/1150911939472523315
* robtfm's branch: https://discord.com/channels/691052431525675048/743663924229963868/1150933597977591809

**Pull request merged**